driver-mapping/src/main/java/com/datastax/driver/mapping/PropertyMapper.java

```diff
@@ -0,0 +1,196 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.mapping;
+
+import com.datastax.driver.core.Metadata;
+import com.datastax.driver.core.TypeCodec;
+import com.datastax.driver.mapping.annotations.*;
+import com.google.common.reflect.TypeToken;
+
+import java.beans.PropertyDescriptor;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Maps a Java bean property to a table column or a UDT field.
+ * <p>
+ * Properties can be either accessed through getter and setter pairs,
+ * or by direct field access, depending on what is available in the
+ * entity/UDT class.
+ */
+class PropertyMapper {
+
+    private final String propertyName;
+    final String alias;
+    final String columnName;
+    final TypeToken<Object> javaType;
+    final TypeCodec<Object> customCodec;
+    final int position;
+
+    private final Field field;
+    private final Method getter;
+    private final Method setter;
+    private final Map<Class<? extends Annotation>, Annotation> annotations;
+
+    PropertyMapper(Class<?> baseClass, String propertyName, String alias, Field field, PropertyDescriptor property) {
```

Comment:
From what I've seen in the code it's never possible for both `Field` and `PropertyDescriptor` to each be defined or both be null, why not have two separate constructors, or even 2 separate subtypes?  I see in the getters/setters a lot of null checking, where if you had two different types you can have separate implementations.